test/adapter_test.rb

```diff
@@ -0,0 +1,68 @@
+require 'test_helper'
+
+class TestSemianAdapter < Minitest::Test
+  def setup
+    Semian.unregister_all_resources
+    # Consumers registered in other test files must be cleared
+    Semian.reset!
+  end
+
+  def teardown
+    Semian.unregister_all_resources
+  end
+
+  def test_adapter_registers_consumer
+    assert_empty(Semian.resources)
+    assert_empty(Semian.consumers)
+    client = Semian::AdapterTestClient.new(quota: 0.5)
+    resource = client.semian_resource
+    assert_equal(resource, Semian.resources[client.semian_identifier])
+    assert_equal(client, Semian.consumers[client.semian_identifier].first)
+  end
+
+  def test_unregister_all_resources
+    client = Semian::AdapterTestClient.new(quota: 0.5)
+    resource = client.semian_resource
+    assert_equal(resource, Semian.resources[client.semian_identifier])
+    assert_equal(client, Semian.consumers[client.semian_identifier].first)
+
+    # need to disable GC to ensure client weak reference is alive for assertion below
+    without_gc do
+      assert_equal(resource, client.semian_resource)
+      Semian.unregister_all_resources
+
+      assert_empty(Semian.resources)
+      assert_empty(Semian.consumers)
+
+      # The first call to client.semian_resource after unregistering all resources,
+      # should return a *different* (new) resource.
+      refute_equal(resource, client.semian_resource)
+    end
+  end
+
+  def test_consumer_registration_does_not_prevent_gc
+    client = Semian::AdapterTestClient.new(quota: 0.5)
+    client.semian_resource
+    identifier = client.semian_identifier
+
+    client = nil
```

Comment:
@fw42 this is to force the garbage collection (by dropping the only strong reference). 

The assert_nil is a hack to prevent rubocop from complaining about the "useless variable assignment".

Alternatively, I suppose I could just whitelist this with rubocop.